File: 补充示例/arma_grid.py
from statsmodels.tsa.arima.model import ARIMA
from statsmodels.stats.diagnostic import acorr_ljungbox
import pandas as pd, numpy as np

# 在配套代码根目录运行；对数差分，不乘100
macro = pd.read_csv("data/processed/wdi_china_macro_1960_2024.csv")
macro = macro[(macro.year>=1978) & (macro.gdp_constant_2015_usd>0)]
macro = macro.sort_values("year")
gdp_growth = np.diff(np.log(macro.gdp_constant_2015_usd.to_numpy()))
# 比较三个规格的 AIC
results = {}
for p, q in [(1,0),(1,1),(2,1)]:
    res = ARIMA(gdp_growth, order=(p,0,q)).fit(
        method_kwargs={"maxiter":1000})
    assert res.mle_retvals["converged"]
    results[(p,q)] = res.aic
print(pd.Series(results, name="AIC"))

# 固定 d=0，在相同候选范围内按 AICc 选择
fits = {}
for p in range(3):
    for q in range(3):
        m = ARIMA(gdp_growth, order=(p,0,q), trend="c").fit(
            method_kwargs={"maxiter":1000})
        if not m.mle_retvals["converged"]:
            raise RuntimeError(f"ARMA({p},{q}) 未收敛，需检查优化")
        fits[(p,q)] = m
order = min(fits, key=lambda k: fits[k].aicc)
m_auto = fits[order]
print(order, m_auto.summary())

# 柳恩--博克斯（Ljung-Box）残差检验
lb = acorr_ljungbox(m_auto.resid, lags=[10],
                   model_df=sum(order), return_df=True)
print(lb)
pd.DataFrame([{"order":f"{p} {q}","AICc":m.aicc} for (p,q),m in fits.items()]).to_csv("补充示例/arma_Python.csv",index=False)

File: 补充示例/arma_grid.R
png("补充示例/arma_R.png",width=1000,height=700)
library(forecast)
library(tseries)

# 在配套代码根目录运行；使用与正文相同的历史快照
macro <- read.csv("data/processed/wdi_china_macro_1960_2024.csv",
                  fileEncoding="UTF-8-BOM")
macro <- macro[macro$year>=1978 & macro$gdp_constant_2015_usd>0, ]
macro <- macro[order(macro$year), ]
gdp_growth <- diff(log(macro$gdp_constant_2015_usd))
ts_growth <- ts(gdp_growth, start=1979, frequency=1)

# 手动估计 ARMA(1,0), ARMA(1,1), ARMA(2,1)
m10 <- Arima(ts_growth, order = c(1, 0, 0))
m11 <- Arima(ts_growth, order = c(1, 0, 1))
m21 <- Arima(ts_growth, order = c(2, 0, 1))
cat("AIC: AR(1)=", AIC(m10), " ARMA(1,1)=", AIC(m11),
    " ARMA(2,1)=", AIC(m21), "\n")

# 自动选择
fits <- list()
for (p in 0:2) for (q in 0:2) {
  fits[[paste(p,q)]] <- Arima(ts_growth, order=c(p,0,q),
                             include.mean=TRUE, method="ML")
}
stopifnot(all(sapply(fits, function(m) m$code == 0)))
m_auto <- fits[[which.min(sapply(fits, function(m) m$aicc))]]
summary(m_auto)

# 柳恩--博克斯（Ljung-Box）残差检验（前 10 滞后）
Box.test(residuals(m_auto), lag=10, type="Ljung-Box",
         fitdf=sum(arimaorder(m_auto)[c(1,3)]))
# 残差 ACF 图
acf(residuals(m_auto), main = "ARMA 残差 ACF")
dev.off()
write.csv(data.frame(order=names(fits),AICc=sapply(fits,function(m)m$aicc)),"补充示例/arma_R.csv",row.names=FALSE)
